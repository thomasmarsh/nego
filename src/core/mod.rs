pub mod bitboard;
pub mod coord;
pub mod error;
pub mod game;
pub mod r#move;
pub mod move_tab;
pub mod orientation;
pub mod pieces;
pub mod ray;
pub mod square;
pub mod zobrist;
