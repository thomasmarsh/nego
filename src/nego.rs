mod ai;
mod core;

#[macro_use]
extern crate log;

fn main() -> rustyline::Result<()> {
    pretty_env_logger::init();

    info!("initializing ray LUT");
    core::ray::Rays::build_lut();

    // cli()

    //dem();
    ai::demo_minimax();
    Ok(())
}
