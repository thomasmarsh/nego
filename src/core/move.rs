use crate::core::{
    bitboard::BitBoard,
    coord::{X, Y},
    error::Error,
    move_tab::LUTEntry,
    orientation::Orientation,
    pieces::{PieceId, PieceTypeId},
    square::Square,
};

use serde::Serialize;
use std::fmt;
use std::str::FromStr;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize)]
pub struct Move(u16);

impl Move {
    #[inline]
    pub fn new(piece: PieceId, entry: LUTEntry) -> Move {
        let p = (piece as u16) & 0b1111;
        let e = (entry.0 as u16) << 4;
        Move(e | p)
    }

    #[inline]
    pub fn get_raw_value(self) -> u16 {
        self.0
    }

    #[inline]
    pub fn get_piece(self) -> PieceId {
        PieceId::from_index(self.0 & 0b1111).unwrap()
    }

    pub fn get_lut_entry(self) -> LUTEntry {
        LUTEntry(self.0.wrapping_shr(4) as usize)
    }

    #[inline]
    pub fn position(&self) -> Square {
        self.get_lut_entry().position()
    }

    #[inline]
    pub fn orientation(&self) -> Orientation {
        self.get_lut_entry().orientation()
    }

    #[inline]
    pub fn mask(&self) -> BitBoard {
        self.get_lut_entry().mask()
    }

    // First set bit (self.position may be empty on kunoji types)
    #[inline]
    pub fn to_square(self) -> Square {
        self.mask().to_square()
    }

    #[inline]
    pub fn gaze(&self) -> BitBoard {
        self.get_lut_entry().gaze()
    }

    pub fn notation(&self) -> String {
        format!(
            "{}:{}{:?}",
            self.get_piece().piece_type_id().notation(),
            self.position().to_string().to_uppercase(),
            self.orientation()
        )
        .to_string()
    }

    pub fn parse(s: &str) -> Result<Move, Error> {
        if s.len() != 8 {
            return Err(Error::InvalidFormat);
        }
        if s.as_bytes()[4] != b':' {
            return Err(Error::InvalidFormat);
        }

        fn parse_piece_type(piece_str: &str) -> Result<PieceTypeId, Error> {
            match piece_str {
                "BOS" => Ok(PieceTypeId::Boss),
                "MAM" => Ok(PieceTypeId::Mame),
                "NOB" => Ok(PieceTypeId::Nobi),
                "KB1" => Ok(PieceTypeId::Koubaku1),
                "KB2" => Ok(PieceTypeId::Koubaku2),
                "KB3" => Ok(PieceTypeId::Koubaku3),
                "KJ1" => Ok(PieceTypeId::Kunoji1),
                "KJ2" => Ok(PieceTypeId::Kunoji2),
                "KJ3" => Ok(PieceTypeId::Kunoji3),
                "KJ4" => Ok(PieceTypeId::Kunoji4),
                _ => Err(Error::InvalidPiece),
            }
        }

        let piece_type = parse_piece_type(&s[0..4])?;
        let x = X::from_str(&s.as_bytes()[5].to_string())?;
        let y = Y::from_str(&s.as_bytes()[6].to_string())?;
        let position = Square::make_square(x, y);
        let orientation = Orientation::from_str(&s.as_bytes()[7].to_string())?;

        let piece = match piece_type {
            PieceTypeId::Boss => PieceId::Boss,
            PieceTypeId::Mame => PieceId::Mame,
            PieceTypeId::Nobi => PieceId::Nobi,
            PieceTypeId::Koubaku1 => PieceId::Koubaku1,
            PieceTypeId::Koubaku2 => PieceId::Koubaku2,
            PieceTypeId::Koubaku3 => PieceId::Koubaku3a, // may need reindexing
            PieceTypeId::Kunoji1 => PieceId::Kunoji1a,   // may need reindexing
            PieceTypeId::Kunoji2 => PieceId::Kunoji2,
            PieceTypeId::Kunoji3 => PieceId::Kunoji3,
            PieceTypeId::Kunoji4 => PieceId::Kunoji4,
        };

        let entry =
            LUTEntry::lookup(piece_type, position, orientation).ok_or(Error::LUTEntryNotFound)?;
        Ok(Move::new(piece, entry))
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.notation())
    }
}

impl FromStr for Move {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Move::parse(s)
    }
}

pub trait MoveVisitor {
    fn visit(&mut self, m: Move);

    #[inline]
    fn bailout(&self) -> bool {
        false
    }
}

pub struct HasMoves(pub bool);

impl MoveVisitor for HasMoves {
    #[inline]
    fn visit(&mut self, _: Move) {
        self.0 = true;
    }

    #[inline]
    fn bailout(&self) -> bool {
        self.0
    }
}

pub struct MoveCounter(pub usize);

impl MoveVisitor for MoveCounter {
    #[inline]
    fn visit(&mut self, _: Move) {
        self.0 += 1;
    }
}

pub struct MoveAccumulator(pub Vec<Move>);

impl MoveAccumulator {
    #[inline]
    pub fn new() -> Self {
        Self(Vec::with_capacity(1720 >> 1))
    }
}

impl MoveVisitor for MoveAccumulator {
    #[inline]
    fn visit(&mut self, m: Move) {
        self.0.push(m);
    }
}
