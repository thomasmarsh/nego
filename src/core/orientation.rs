use crate::core::{error::Error, square::Square};

use std::str::FromStr;

#[derive(PartialEq, Clone, Copy, Debug, Eq)]
pub enum Orientation {
    S, // S, 0 (reference orientation of pieces)
    W, // E, 90
    N, // N, 180
    E, // W, 270
}

pub const NUM_ORIENTATIONS: usize = 4;

pub const ALL_ORIENTATIONS: [Orientation; 4] = [
    Orientation::S,
    Orientation::W,
    Orientation::N,
    Orientation::E,
];

impl Orientation {
    #[inline]
    pub fn at_limit(&self, square: Square) -> bool {
        use Orientation::*;
        match self {
            S => square.is_bottom_edge(),
            W => square.is_left_edge(),
            N => square.is_top_edge(),
            E => square.is_right_edge(),
        }
    }

    #[inline]
    pub fn opposite(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => N,
            W => E,
            N => S,
            E => W,
        }
    }

    // if "facing" one direction, what direction is to the right?
    #[inline]
    pub fn right(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => W,
            W => N,
            N => E,
            E => S,
        }
    }

    // if "facing" one direction, what direction is to the left?
    #[inline]
    pub fn left(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => E,
            W => S,
            N => W,
            E => N,
        }
    }

    #[inline]
    pub fn from_index(i: u8) -> Orientation {
        match i {
            0 => Orientation::S,
            1 => Orientation::W,
            2 => Orientation::N,
            3 => Orientation::E,
            _ => unreachable!(),
        }
    }
}

impl FromStr for Orientation {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            return Err(Error::InvalidRank);
        }
        match s.chars().next().unwrap() {
            'S' => Ok(Orientation::S),
            'W' => Ok(Orientation::W),
            'N' => Ok(Orientation::N),
            'E' => Ok(Orientation::E),
            _ => Err(Error::InvalidOrientation),
        }
    }
}
