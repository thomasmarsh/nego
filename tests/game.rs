use nego::core::{
    game::State, move_tab::LUTEntry, orientation::Orientation, pieces::PieceId, r#move::Move,
    square::Square,
};

fn mk_move(piece: PieceId, pos: (u8, u8), orientation: Orientation) -> Move {
    let entry = LUTEntry::lookup(
        piece.piece_type_id(),
        Square::from_indices(pos.0 as usize, pos.1 as usize),
        orientation,
    )
    .unwrap();
    Move::new(piece, entry)
}

fn play_moves(state: &mut State, moves: &[(PieceId, (u8, u8), Orientation)]) {
    moves.iter().for_each(|(a, b, c)| {
        state.place(mk_move(*a, *b, *c));
        state.current = state.current.next();
    });
}

#[test]
fn no_capture_boss() {
    use Orientation::*;
    use PieceId::*;

    let mut state = State::new();
    let first_boss = mk_move(Boss, (0, 0), S);
    state.place(first_boss);
    state.current = state.current.next();
    state.dump();
    let occupied = state.board.black.occupied;

    play_moves(
        &mut state,
        &[
            (Boss, (0, 2), S), // W
            (Mame, (5, 5), S), // B
        ],
    );

    println!("before placement:");
    state.dump();
    state.place(mk_move(Nobi, (2, 0), W)); // W
    println!("after placement:");
    state.dump();

    assert!(state.board.black.occupied & occupied == occupied);
    assert_eq!(state.board.black.move_list[0].get_piece(), Boss);
}

#[test]
fn snapshot() {
    use rand::{seq::SliceRandom, SeedableRng};
    let mut rng = rand::rngs::StdRng::seed_from_u64(123);

    for _ in 0..20 {
        let mut state = State::new();
        let mut moves = Vec::new();
        let mut history = Vec::new();
        loop {
            history.push(state.clone());
            moves.truncate(0);
            state.get_moves(&mut moves);
            if moves.is_empty() {
                break;
            }
            let m = *moves.choose(&mut rng).unwrap();
            state.apply(m);
        }

        insta::assert_debug_snapshot!(history);
    }
}
