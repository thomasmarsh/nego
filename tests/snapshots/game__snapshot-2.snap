---
source: tests/game.rs
expression: history
---
[
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                0,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 0,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                ],
                occupied: BitBoard(
                    3311419785216,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419785216,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 1070485894089486513,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                ],
                occupied: BitBoard(
                    3311419785216,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                ],
                occupied: BitBoard(
                    49344,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 8338037931911035167,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4078,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                ],
                occupied: BitBoard(
                    3311419785984,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                ],
                occupied: BitBoard(
                    49344,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 13538484330490851644,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4078,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                ],
                occupied: BitBoard(
                    3311419785984,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4090,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                ],
                occupied: BitBoard(
                    2015424,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 1313014549381515442,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3054,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                ],
                occupied: BitBoard(
                    4138200990464,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4090,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                ],
                occupied: BitBoard(
                    2015424,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 16013878874391626450,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3054,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                ],
                occupied: BitBoard(
                    4138200990464,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4088,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                ],
                occupied: BitBoard(
                    17592188059840,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 13839231501724613325,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3022,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                ],
                occupied: BitBoard(
                    4138207281920,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4088,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                ],
                occupied: BitBoard(
                    17592188059840,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 8454591529032596242,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3022,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                ],
                occupied: BitBoard(
                    4138207281920,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4080,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                ],
                occupied: BitBoard(
                    17592389386432,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 18289398563564160102,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3020,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                ],
                occupied: BitBoard(
                    4139281023744,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4080,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                ],
                occupied: BitBoard(
                    17592389386432,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 1419346379863201087,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3020,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                ],
                occupied: BitBoard(
                    4139281023744,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4064,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                ],
                occupied: BitBoard(
                    6772991830442176,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 15607861608353005342,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3016,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                ],
                occupied: BitBoard(
                    4139281023774,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4064,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                ],
                occupied: BitBoard(
                    6772991830442176,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 9069302887064932222,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3016,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                ],
                occupied: BitBoard(
                    4139281023774,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    2016,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                ],
                occupied: BitBoard(
                    43012895081939136,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 13793520176559011402,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2888,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                ],
                occupied: BitBoard(
                    27060921417335582,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    2016,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                ],
                occupied: BitBoard(
                    43012895081939136,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 9390834442801454530,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2888,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                ],
                occupied: BitBoard(
                    27060921417335582,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1504,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                    Move(
                        19129,
                    ),
                ],
                occupied: BitBoard(
                    43021794254176448,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 3405496757959002133,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2824,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                    Move(
                        14310,
                    ),
                ],
                occupied: BitBoard(
                    459406485644903198,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1504,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                    Move(
                        19129,
                    ),
                ],
                occupied: BitBoard(
                    43021794254176448,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 1896738041639290650,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2824,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                    Move(
                        14310,
                    ),
                ],
                occupied: BitBoard(
                    459406485644903198,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1472,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                    Move(
                        19129,
                    ),
                    Move(
                        13765,
                    ),
                ],
                occupied: BitBoard(
                    43021794254182592,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 14811609324649962594,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2816,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                    Move(
                        14310,
                    ),
                    Move(
                        7779,
                    ),
                ],
                occupied: BitBoard(
                    459406623620727582,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1472,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                    Move(
                        19129,
                    ),
                    Move(
                        13765,
                    ),
                ],
                occupied: BitBoard(
                    43021794254182592,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 2954169068352225697,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2816,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                    Move(
                        14310,
                    ),
                    Move(
                        7779,
                    ),
                ],
                occupied: BitBoard(
                    459406623620727582,
                ),
                owned: BitBoard(
                    799,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1408,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                    Move(
                        19129,
                    ),
                    Move(
                        13765,
                    ),
                    Move(
                        14534,
                    ),
                ],
                occupied: BitBoard(
                    43021794271025344,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 16741992486415781195,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2304,
                ),
                move_list: [
                    Move(
                        432,
                    ),
                    Move(
                        9844,
                    ),
                    Move(
                        22538,
                    ),
                    Move(
                        15445,
                    ),
                    Move(
                        3889,
                    ),
                    Move(
                        4370,
                    ),
                    Move(
                        17815,
                    ),
                    Move(
                        14310,
                    ),
                    Move(
                        7779,
                    ),
                    Move(
                        20441,
                    ),
                ],
                occupied: BitBoard(
                    461099871527502622,
                ),
                owned: BitBoard(
                    506381192568767263,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    1408,
                ),
                move_list: [
                    Move(
                        96,
                    ),
                    Move(
                        4530,
                    ),
                    Move(
                        4081,
                    ),
                    Move(
                        6963,
                    ),
                    Move(
                        10452,
                    ),
                    Move(
                        25899,
                    ),
                    Move(
                        19129,
                    ),
                    Move(
                        13765,
                    ),
                    Move(
                        14534,
                    ),
                ],
                occupied: BitBoard(
                    43021794271025344,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                3311419834560,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 2109837930653034743,
        capture_flag: false,
    },
]
